File: Android/Sources/UserControlLibraries/GaodeMaps/src/main/java/com/genexus/android/controls/maps/gaode/GxMapViewCamera.java
package com.genexus.android.controls.maps.gaode;

import android.graphics.Point;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.amap.api.maps.AMap;
import com.amap.api.maps.CameraUpdate;
import com.amap.api.maps.CameraUpdateFactory;
import com.amap.api.maps.MapView;
import com.amap.api.maps.model.CameraPosition;
import com.amap.api.maps.model.LatLng;
import com.genexus.android.core.base.services.Services;
import com.genexus.android.maps.GxMapViewBase;
import com.genexus.android.maps.MapViewCamera;
import com.genexus.android.core.controls.maps.common.IMapLocation;
import com.genexus.android.maps.MapItemViewHelper;

class GxMapViewCamera extends MapViewCamera<CameraUpdate, MapLocation, MapLocationBounds, GxMapViewItem, GxMapViewData> {
	private final AMap mMap;
	private final MapView mMapView;

	public GxMapViewCamera(AMap map, MapView mapView) {
		mMap = map;
		mMapView = mapView;
	}

	@Override
	public void updateCamera(@NonNull MapLocationBounds bounds, int padding, int duration) {
		CameraUpdate update = CameraUpdateFactory.newLatLngBounds(bounds.getLatLngBounds(), padding);
		updateCamera(update, duration, null);
	}

	@Override
	public void updateCamera(@NonNull IMapLocation location, int zoom, int duration, GxMapViewBase.OnMarkerCameraUpdateCallback callback) {
		updateCamera(getCameraUpdate(location, zoom), duration, callback);
	}

	@Override
	public void updateCamera(@NonNull IMapLocation location, int zoom) {
		updateCamera(getCameraUpdate(location, zoom), 0, null);
	}

	@Override
	public void updateCamera(CameraUpdate cameraUpdate, int duration, GxMapViewBase.OnMarkerCameraUpdateCallback callback) {
		if (mMap != null) {
			try {
				if (duration == 0)
					mMap.animateCamera(cameraUpdate, getCallback(callback));
				else
					mMap.animateCamera(cameraUpdate, duration, getCallback(callback));

				return;
			} catch (IllegalStateException e) {
				Services.Log.error(e);
			}
		}

		setPendingUpdate(cameraUpdate);
	}

	@Nullable
	@Override
	public Projection<MapLocation, MapLocationBounds> getProjection(MapLocation initialLocation) {
		if (mMap == null)
			return null;

		com.amap.api.maps.Projection projection = mMap.getProjection();
		CameraPosition cameraPosition = mMap.getCameraPosition();

		MapLocation fromScreenLocation = new MapLocation(0D, 0D);
		if (initialLocation != null) {
			Point centerPoint = projection.toScreenLocation(new LatLng(initialLocation.getLatitude(), initialLocation.getLatitude()));
			centerPoint.y -= (int) (mMapView.getHeight() * MapItemViewHelper.MARKER_INFO_WINDOW_OFF_CENTER_FACTOR);
			fromScreenLocation = new MapLocation(projection.fromScreenLocation(centerPoint));
		}

		MapLocation center = new MapLocation(cameraPosition.target.latitude, cameraPosition.target.longitude);
		MapLocationBounds latLngBounds = new MapLocationBounds(projection.getVisibleRegion().latLngBounds);
		int zoom = (int) cameraPosition.zoom;

		return new Projection<>(latLngBounds, center, fromScreenLocation, zoom);
	}

	private CameraUpdate getCameraUpdate(@NonNull IMapLocation location, int zoom) {
		LatLng latLng = new LatLng(location.getLatitude(), location.getLongitude());
		CameraUpdate update;
		if (zoom > 0)
			update = CameraUpdateFactory.newLatLngZoom(latLng, zoom);
		else
			update = CameraUpdateFactory.newLatLng(latLng);

		return update;
	}

	private AMap.CancelableCallback getCallback(GxMapViewBase.OnMarkerCameraUpdateCallback callback) {
		return new AMap.CancelableCallback() {
			@Override
			public void onFinish() {
				if (callback != null)
					callback.onFinish();
			}

			@Override
			public void onCancel() {
				if (callback != null)
					callback.onCancel();
			}
		};
	}
}

File: Android/Sources/UserControlLibraries/HMSMapKitLib/src/main/java/com/genexus/android/controls/maps/hms/GxMapViewCamera.kt
package com.genexus.android.controls.maps.hms

import android.graphics.Point
import com.genexus.android.core.base.services.Services
import com.genexus.android.core.controls.maps.common.IMapLocation
import com.genexus.android.maps.GxMapViewBase
import com.genexus.android.maps.MapItemViewHelper
import com.genexus.android.maps.MapViewCamera
import com.huawei.hms.maps.CameraUpdate
import com.huawei.hms.maps.CameraUpdateFactory
import com.huawei.hms.maps.HuaweiMap
import com.huawei.hms.maps.MapView
import com.huawei.hms.maps.model.LatLng

internal class GxMapViewCamera(private val map: HuaweiMap?, private val mapView: MapView) :
    MapViewCamera<CameraUpdate, MapLocation, MapLocationBounds, GxMapViewItem, GxMapViewData>() {

    override fun updateCamera(bounds: MapLocationBounds, padding: Int, duration: Int) {
        val update = CameraUpdateFactory.newLatLngBounds(bounds.latLngBounds, padding)
        updateCamera(update, duration, null)
    }

    override fun updateCamera(
        location: IMapLocation,
        zoom: Int,
        duration: Int,
        callback: GxMapViewBase.OnMarkerCameraUpdateCallback?
    ) {
        updateCamera(getCameraUpdate(location, zoom), duration, callback)
    }

    override fun updateCamera(location: IMapLocation, zoom: Int) {
        updateCamera(getCameraUpdate(location, zoom), 0, null)
    }

    override fun updateCamera(
        cameraUpdate: CameraUpdate,
        duration: Int,
        callback: GxMapViewBase.OnMarkerCameraUpdateCallback?
    ) {
        if (map != null) {
            try {
                if (duration == 0)
                    map.animateCamera(cameraUpdate, getCallback(callback))
                else
                    map.animateCamera(cameraUpdate, duration, getCallback(callback))
                return
            } catch (e: IllegalStateException) {
                Services.Log.error(e)
            }
        }
        pendingUpdate = cameraUpdate
    }

    override fun getProjection(initialLocation: MapLocation?): Projection<MapLocation, MapLocationBounds>? {
        if (map == null) return null
        val projection = map.projection
        val cameraPosition = map.cameraPosition
        var fromScreenLocation = MapLocation(0.0, 0.0)
        if (initialLocation != null) {
            val centerPoint: Point = projection.toScreenLocation(LatLng(initialLocation.latitude, initialLocation.longitude))
            centerPoint.y -= (mapView.height * MapItemViewHelper.MARKER_INFO_WINDOW_OFF_CENTER_FACTOR).toInt()
            fromScreenLocation = MapLocation(projection.fromScreenLocation(centerPoint))
        }
        val center = MapLocation(cameraPosition.target.latitude, cameraPosition.target.longitude)
        val latLngBounds = MapLocationBounds(projection.visibleRegion.latLngBounds)
        val zoom = cameraPosition.zoom.toInt()
        return Projection(latLngBounds, center, fromScreenLocation, zoom)
    }

    private fun getCameraUpdate(location: IMapLocation, zoom: Int): CameraUpdate {
        val latLng = LatLng(location.latitude, location.longitude)
        return if (zoom > 0)
            CameraUpdateFactory.newLatLngZoom(latLng, zoom.toFloat())
        else
            CameraUpdateFactory.newLatLng(latLng)
    }

    private fun getCallback(callback: GxMapViewBase.OnMarkerCameraUpdateCallback?): HuaweiMap.CancelableCallback {
        return object : HuaweiMap.CancelableCallback {
            override fun onFinish() {
                callback?.onFinish()
            }

            override fun onCancel() {
                callback?.onCancel()
            }
        }
    }
}
